File: programcreek/js/01/01-09-two-sum.js


function twoSum(list, target) {
    for (var i = 0; i < list.length - 1; i ++){
        for (var j = i + 1; j < list.length ; j++) {
            if (list[i] + list[j] == target) {
                return [i, j];
            }
        }
    }
    return null;
}


function twoSumWithHash(list, target) {
    var table = {}
    for (var i = 0; i < list.length; i ++) {
        if (typeof table[ list[i] ] != 'undefined' ) {
            return [ table[list[i]] , i ]
        }
        else {
            table[ target - list[i] ] = i;
        }
    }
    return null;
}


function twoSumWithSortedInput(list, target) {
    var start = 0;
    var end = list.length - 1;
    while (start < end) {
        var sum = list[start] + list[end];
        if (sum == target) return [start, end];
        if (sum < target) start ++;
        else end --;
    }
    return null;
}

function TwoSumData(){
    this.table = {}
}

TwoSumData.prototype.add = function(i) {
    if (typeof this.table[i] == 'undefined') {
        this.table[i] = 0;
    }
    this.table[i] += 1
}

TwoSumData.prototype.find = function(target) {
    for (var i in this.table) {
        var reminder = target - i;
        if (typeof this.table[reminder] != 'undefined') {
            if (i == reminder && this.table[reminder] < 2) {
                continue;
            }
            return true;
        }
    }
    return false;
}


console.log(twoSum([2,7,11,15], 22));

console.log(twoSumWithHash([2,7,11,15], 22));

console.log(twoSumWithSortedInput([2,7,11,15,16,19,23,33], 30));


var c = new TwoSumData();
c.add(1);
c.add(3);
c.add(7);
console.log(c.find(4));
console.log(c.find(7));
console.log(c.find(6));

File: programcreek/python/01/01-09-two-sum.py
def twoSum(nums, target):
    for i in range(len(nums) - 1):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]
    return None




def twoSumWithHash(nums, target):
    table = {}
    for i in range(len(nums)):
        reminder = target - nums[i]
        if nums[i] in table:
            return [table[nums[i]], i]
        else:
            table[reminder] = i
    return None



def twoSumWithSortedInput(nums, target):
    start = 0
    end = len(nums) - 1
    while start < end:
        sum = nums[start] + nums[end]
        if sum == target:
            return [start, end]
        if sum > target:
            end -= 1
        else:
            start += 1
    return None





class TwoSumData :
    table = None

    def __init__(self):
        self.table = dict()

    def add(self, i):
        if i not in self.table:
            self.table[i] = 0
        self.table[i] += 1

    def find(self, target):
        for i in self.table:
            reminder = target - i
            if reminder in self.table:
                if i == reminder and self.table[i] < 2:
                    continue
                return True
        return False




print twoSum([2,7,11,15], 18)

print twoSumWithHash([2,7,11,15], 18)

print twoSumWithSortedInput([2,7,11,15,16,19,23,33], 30)


c = TwoSumData()
c.add(1);
c.add(3);
c.add(7);
print c.find(4)
print c.find(7)
print c.find(6)